src/Razor/Microsoft.AspNetCore.Razor.Language/test/IntegrationTests/ComponentCodeGenerationTestBase.cs

```diff
@@ -3152,6 +3152,391 @@ public class MyComponent<TItem> : ComponentBase
+    }
+
+    public class Column<TItem> : ComponentBase
+    {
+        [Parameter] public RenderFragment<TItem> ChildContent { get; set; }
+    }
+}
+"));
+
+            // Act
+            var generated = CompileToCSharp(@"
+<Grid Items=""@(Array.Empty<DateTime>())""><Column>@context.Year</Column></Grid>");
+
+            // Assert
+            AssertDocumentNodeMatchesBaseline(generated.CodeDocument);
+            AssertCSharpDocumentMatchesBaseline(generated.CodeDocument);
+            CompileToAssembly(generated);
+        }
+
+        [Fact]
+        public void ChildComponent_Generic_TypeInference_CascadedWithLambda()
+        {
+            // Arrange
+            AdditionalSyntaxTrees.Add(Parse(@"
+using Microsoft.AspNetCore.Components;
+
+namespace Test
+{
+    [CascadingTypeParameter(nameof(TItem))]
+    public class Grid<TItem> : ComponentBase
+    {
+        [Parameter] public System.Collections.Generic.IEnumerable<TItem> Items { get; set; }
+        [Parameter] public RenderFragment ChildContent { get; set; }
+    }
+
+    public class Column<TItem, TOutput> : ComponentBase
+    {
+        [Parameter] public System.Func<TItem, TOutput> SomeLambda { get; set; }
+    }
+}
+"));
+
+            // Act
+            var generated = CompileToCSharp(@"
+<Grid Items=""@(Array.Empty<DateTime>())""><Column SomeLambda=""@(x => x.Year)"" /></Grid>");
+
+            // Assert
+            AssertDocumentNodeMatchesBaseline(generated.CodeDocument);
+            AssertCSharpDocumentMatchesBaseline(generated.CodeDocument);
+            CompileToAssembly(generated);
+        }
+
+        [Fact]
+        public void ChildComponent_Generic_TypeInference_CascadedWithMultipleTypes()
+        {
+
+            // Arrange
+            AdditionalSyntaxTrees.Add(Parse(@"
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Components;
+
+namespace Test
+{
+    [CascadingTypeParameter(nameof(TKey))]
+    [CascadingTypeParameter(nameof(TValue))]
+    [CascadingTypeParameter(nameof(TOther))]
+    public class Parent<TKey, TValue, TOther> : ComponentBase
+    {
+        [Parameter] public Dictionary<TKey, TValue> Data { get; set; }
+        [Parameter] public TOther Other { get; set; }
+        [Parameter] public RenderFragment ChildContent { get; set; }
+    }
+
+    public class Child<TOther, TValue, TKey, TChildOnly> : ComponentBase
+    {
+        [Parameter] public ICollection<TChildOnly> ChildOnlyItems { get; set; }
+    }
+}
+"));
+
+            // Act
+            var generated = CompileToCSharp(@"
+<Parent Data=""@(new System.Collections.Generic.Dictionary<int, string>())"" Other=""@DateTime.MinValue"">
+    <Child ChildOnlyItems=""@(new[] { 'a', 'b', 'c' })"" />
+</Parent>");
+
+            // Assert
+            AssertDocumentNodeMatchesBaseline(generated.CodeDocument);
+            AssertCSharpDocumentMatchesBaseline(generated.CodeDocument);
+            CompileToAssembly(generated);
+        }
+
```

Comment:
Do we have a test for what happens when a parent and a grand parent cascade the same generic type parameter? 

Who wins? (I believe the parent wins over the GP, but we should capture that behavior in a test)

Reply:
(I'm thinking of stuff like TValue and TItem)

Reply:
Good point - added.